kafka-client-examples/e2e-test/src/main/java/dev/responsive/examples/regression/ResultsComparatorService.java

```diff
@@ -0,0 +1,184 @@
+        var responsive = responsiveBuffered.get(p).iterator();
+        var baseline = baselineBuffered.get(p).iterator();
+
+        while (responsive.hasNext() && baseline.hasNext()) {
+          final var r = responsive.next();
+          final var b = baseline.next();
+
+          if (r.compareTo(b) == 0) {
+            responsive.remove();
+            baseline.remove();
+            matches++;
+          } else if (responsive.hasNext() && baseline.hasNext()) {
+            LOG.error(
+                "Expected to consume records in order, but they were not in order. "
+                    + "Most recent record from responsive is {} and most recent "
+                    + "record from baseline is {}",
+                r.record,
+                b.record
+            );
+          } else {
+            // one of the streams is behind so we'll wait for the next
+            // poll to see if any records come up here
+            break;
+          }
+        }
+      }
+
+      LOG.info(
+          "Matched {} records. Current buffers are: responsive={}, baseline={}",
+          matches,
+          responsiveBuffered.size(),
+          baselineBuffered.size()
+      );
+    }
+  }
+
+  private void pauseResume(final int[] consumedResponsive, final int[] consumedBaseline) {
+    final List<TopicPartition> toPause = new ArrayList<>();
+    final List<TopicPartition> toResume = new ArrayList<>();
+
+    for (int p = 0; p < NUM_PARTITIONS; p++) {
+      final int compare = consumedResponsive[p] - consumedBaseline[p];
+      if (compare > MAX_BUFFER) {
+        LOG.warn(
+            "partition {} on responsive is ahead of baseline by {} records... Pausing baseline.",
+            p,
+            compare
+        );
+        toPause.add(new TopicPartition(resultsTopic(true), p));
+        toResume.add(new TopicPartition(resultsTopic(false), p));
+      } else if (compare < -MAX_BUFFER) {
+        LOG.warn(
+            "partition {} on responsive is behind of baseline by {} records... Pausing responsive.",
+            p,
+            -compare
+        );
+        toPause.add(new TopicPartition(resultsTopic(false), p));
+        toResume.add(new TopicPartition(resultsTopic(true), p));
+      } else {
+        toResume.add(new TopicPartition(resultsTopic(true), p));
+        toResume.add(new TopicPartition(resultsTopic(false), p));
+      }
+    }
+
+    consumer.pause(toPause);
+    consumer.resume(toResume);
+  }
+
+  private static List<TopicPartition> topicPartitions() {
+    return IntStream
+        .range(0, NUM_PARTITIONS)
+        .boxed()
+        .flatMap(p -> Stream.of(
+            new TopicPartition(resultsTopic(true), p),
+            new TopicPartition(resultsTopic(false), p)
+        )).toList();
+  }
+
+  private record Record(ConsumerRecord<String, EnrichedOrder> record)
+      implements Comparable<Record> {
+
+    @Override
+    public int compareTo(final Record o) {
+      return record.value().compareTo(o.record().value());
```

Comment:
this is the logic that compares two records -- we effectively only call this method to compare two output records. It would be good to review and see if we want to compare anything else (e.g. offset, but I don't think offset is necessarily deterministic)